lib/raft/raft.hpp

```diff
@@ -0,0 +1,512 @@
+                    term = msg.term;
+                }
+
+                reply.type = Message::Type::AppendEntriesReply;
+
+                // If the request comes from an older term, discard itj
+                if (msg.term < term) {
+                    // TODO: add my own term
+                    reply.append_entries_reply.success = false;
+                    return true;
+                }
+
+                // If the entry described as previous entry in the message does
+                // not exist, discard this request
+                auto prev_index = msg.append_entries_request.previous_entry_index;
+                auto prev_term = msg.append_entries_request.previous_entry_term;
+                if (prev_index > 0 && prev_term > 0) {
+                    if (!log.find_entry(prev_term, prev_index)) {
+                        reply.append_entries_reply.success = false;
+                        return true;
+                    }
+                }
+
+                // Append the new entries to the log
+                log.merge(msg.append_entries_request.entries,
+                          msg.append_entries_request.count);
+
+                // Update the commit value
+                auto leader_commit = msg.append_entries_request.leader_commit;
+                if (leader_commit > commit_index) {
+                    auto new_index = std::min(leader_commit, log.last_index());
+                    commit_log_entries(commit_index, new_index);
+                    commit_index = new_index;
+                }
+
+                // Checks for conflicting entries, meaning entries that have
+                // the same index as one in the log, but not with the same
+                // term, then solves this conflict.
+
+                reply.append_entries_reply.success = true;
+                reply.append_entries_reply.last_index = log.last_index();
+
+                return true;
+            }
+
+            case Message::Type::AppendEntriesReply: {
+                if (msg.append_entries_reply.success) {
+                    auto last_index = msg.append_entries_reply.last_index;
+                    for (auto i = 0; i < peer_count; i++) {
+                        if (peers[i]->id == msg.from_id) {
+                            peers[i]->match_index = last_index;
+                            peers[i]->next_index = last_index + 1;
+                        }
+                    }
+                    auto new_index = find_safe_index();
+                    commit_log_entries(commit_index, new_index);
+                    commit_index = new_index;
+                } else {
+                    for (auto i = 0; i < peer_count; i++) {
+                        if (peers[i]->id == msg.from_id) {
+                            peers[i]->next_index --;
+                        }
+                    }
+                }
+                break;
+            }
+        }
+
+        return false;
+    }
+
+    void start_election()
```

Comment:
Since all members are public, you could consider making these free functions and inject the class as argument